### PR TITLE
Add tesseract kinetic and melodic docker files

### PR DESCRIPTION
**Pull request by Levi-Armstrong**

### CODE REVIEW

**Comment by haudren** [comment]
Thanks for putting these up!

I have a few remarks:
- I believe you should use `ENV` instead of `ARG` for `DEBIAN_FRONTEND`
- Those two build scripts are almost identical, they would greatly benefit from being a single Dockerfile with an arg to switch between melodic & kinetic
- I would use a multi-stage docker build to copy the build artifacts instead of removing sources after building as they will remain in one of the layers...

Example:
```
FROM ros:melodic-perception as builder

RUN wget && cmake && make && make install

FROM ros:melodic-perception
COPY --from=builder /usr/local /usr/local
```

(Documentation here: https://docs.docker.com/develop/develop-images/multistage-build/ )

**Comment by gavanderhoorn** [comment]
Multi-stage builds are a good idea, but it does take a bit of care to make sure you also install required runtime dependencies @haudren.

Just copying `/usr/local` will most likely not be sufficient.

**Comment by haudren** [comment]
Yes sorry, my example only covers installing your source dependencies. I guess a more complete template is as follows:
```
FROM ros:melodic-perception as builder

RUN apt-get install all my build dependencies
RUN wget && cmake && make && make install

FROM ros:melodic-perception
RUN apt-get install all my runtime dependencies
COPY --from=builder /usr/local /usr/local
```

**Comment by gavanderhoorn** [comment]
> ```dockerfile
> RUN apt-get install all my build dependencies
> ```

yes, and here is the part where it gets more difficult -- and typically fully manual :)

**Comment by Levi-Armstrong** [comment]
> Those two build scripts are almost identical, they would greatly benefit from being a single Dockerfile with an arg to switch between melodic & kinetic

Do you have an example of doing what you suggested?

File: tesseract/kinetic/Dockerfile
Comment (gavanderhoorn) [inline tesseract/kinetic/Dockerfile]: `ARG` is actually better :slightly_smiling_face: 

`ENV` ends up in the resulting image, which causes *all* future invocations of `apt` to be noninteractive, which is not good.

**Comment by gavanderhoorn** [comment]
```dockerfile
ARG TARGET_ROS_VERSION=melodic
ARG DEBIAN_FRONTEND=noninteractive

FROM osrf/ros:$TARGET_ROS_VERSION-desktop

# redeclare as FROM resets everything
ARG TARGET_ROS_VERSION=melodic
```
etc

**Comment by haudren** [comment]
Simple example:
```dockerfile
ARG ROSDISTRO=melodic
FROM ros:$ROSDISTRO
ARG ROSDISTRO
RUN apt-get update && apt-get install -y ros-$ROSDISTRO-pr2-common
```

Then you can do:
``` 
docker build
```
To generate the melodic version and the following for kinetic:
```
docker build . --build-arg ROSDISTRO=kinetic
```

**Comment by haudren** [comment]
Oh and excellent remark for the `ARG`, I always assumed it did not set the environment variable...

**Comment by Levi-Armstrong** [comment]
@gavanderhoorn Do you mind taking a look?

**Comment by gavanderhoorn** [comment]
I've only looked at the Tesseract `Dockerfile`. Some high-level comments.

Convert `RUN apt .. install ..` to:

```Dockerfile
RUN apt-get update \
 && apt-get install -y --no-install-recommends \
      ...
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/*
```

Make sure to remove build directories in the same `RUN` in which you create and build everything.

Minimise the nr of RUN statements in general.

Unless you want to keep the sources around of the stuff you build, delete all of that as well. Make sure to do that in the same layer again (ie: same `RUN`).

If you want to make things really efficient, consider using multi-stage `Dockerfile`s.

**Comment by Levi-Armstrong** [comment]
@gavanderhoorn I believe I made the recommended changes.

**Comment by Levi-Armstrong** [comment]
@gavanderhoorn How do things look now?

**Comment by Levi-Armstrong** [comment]
@gavanderhoorn It this good to merge? I believe CI will pass after it is merged because it needs the updated docker that gets generated from this repository.

File: ros-core-nvidia/kinetic/Dockerfile
Comment (gavanderhoorn) [inline ros-core-nvidia/kinetic/Dockerfile]: I guess this is out-of-scope for this PR, but we should really update this.

File: ros-core-nvidia/kinetic/Dockerfile
Comment (gavanderhoorn) [inline ros-core-nvidia/kinetic/Dockerfile]: Does this end up doing the same thing as `ENV` (ie: hard-code this for every Docker image build on top of this one)?

If that's the case, I would suggest to use the `ARG`  approach here as well (https://github.com/ros-industrial/docker/pull/30#pullrequestreview-346650656).

File: tesseract/Dockerfile
Comment (gavanderhoorn) [inline tesseract/Dockerfile]: Nit, but: best practice is to keep this alphabetically ordered.

Makes finding a particular package in this list that much easier.

Comment (gavanderhoorn) [reply tesseract/Dockerfile]: Additionally: not something you need to change, but the suggestion in https://github.com/ros-industrial/docker/pull/30#issuecomment-601842948 was not necessarily to `apt install` all packages in one `RUN`. Merely to use the suggested workflow / sequence of statements for all `RUN` lines which installed software using `apt`.

It's ok to leave this now as-is, but it could be beneficial for layer reuse to split things out and only `install` them when/where they are needed.

File: tesseract/Dockerfile
Comment (gavanderhoorn) [inline tesseract/Dockerfile]: For future `Dockerfile`s: use [WORKDIR](https://docs.docker.com/engine/reference/builder/#workdir). It combines these two steps.

File: tesseract/Dockerfile
Comment (gavanderhoorn) [inline tesseract/Dockerfile:29]: Would you perhaps want to lock this clone to a certain commit?

Also fine if you want to track `master`, just wanted to make sure.

File: tesseract/Dockerfile
Comment (gavanderhoorn) [inline tesseract/Dockerfile]: Something to consider: did you split the cloning and the build into two `RUN`s deliberately? Restarted `docker build`s do benefit from this, as they don't need to repeat the `clone` again, so this may be something to keep.

File: tesseract/Dockerfile
Comment (gavanderhoorn) [inline tesseract/Dockerfile]: Is this really necessary?

Are dependent targets not linking to libs with absolute paths?

IIRC, `/usr/local/lib` should already be on the standard linker search path.

**Review by gavanderhoorn — APPROVED** [review]
Some minor comments and one minor change requested.

Other than, this looks ok.

I would suggest to remove the Lunar and Indigo `Dockerfile`s, but perhaps we're keeping those for archival reasons.

File: ros-core-nvidia/kinetic/Dockerfile
Comment (Levi-Armstrong) [reply ros-core-nvidia/kinetic/Dockerfile]: I can add my email; do you want to be added?

File: tesseract/Dockerfile
Comment (Levi-Armstrong) [reply tesseract/Dockerfile]: I updated the to follow the alphabetically ordered.

Comment (Levi-Armstrong) [reply tesseract/Dockerfile]: >Additionally: not something you need to change, but the suggestion in #30 (comment) was not necessarily to apt install all packages in one RUN. Merely to use the suggested workflow / sequence of statements for all RUN lines which installed software using apt.

Are you recommending what is below?
```
RUN apt-get update
RUN apt-get install -y --no-install-recommends clang && apt-get clean && rm -rf /var/lib/apt/lists/*
RUN apt-get install -y --no-install-recommends clang-tidy && apt-get clean && rm -rf /var/lib/apt/lists/*
```

File: tesseract/Dockerfile
Comment (Levi-Armstrong) [reply tesseract/Dockerfile]: Done

File: tesseract/Dockerfile
Comment (Levi-Armstrong) [reply tesseract/Dockerfile:29]: Leaving it will master may allow to catch breaking changes in Bullet.

File: tesseract/Dockerfile
Comment (Levi-Armstrong) [reply tesseract/Dockerfile]: Not sure why I broke up things. Are you recommending making them one line? If that is the case how do you leverage WORKDIR? 

```
RUN git clone --branch master https://github.com/bulletphysics/bullet3.git
WORKDIR /bullet3_build
RUN cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-w -std=c++14" -DCMAKE_C_FLAGS="-w" -DBUILD_UNIT_TESTS=OFF -DBUILD_SHARED_LIBS=ON -DBULLET2_MULTITHREADING=OFF -DUSE_DOUBLE_PRECISION=ON ../bullet3 && \
    make -j8 && \
    make install && \
    cd .. && \
    rm -rf /bullet3 /bullet3_build 
```

File: tesseract/Dockerfile
Comment (Levi-Armstrong) [reply tesseract/Dockerfile]: I don't believe it is a default path on Ubuntu based on [this](https://askubuntu.com/questions/165027/why-cant-my-program-find-the-libraries-in-usr-local-lib) discussion.

File: ros-core-nvidia/kinetic/Dockerfile
Comment (Levi-Armstrong) [reply ros-core-nvidia/kinetic/Dockerfile]: I do not know. I copied this from the other dockers.

File: tesseract/Dockerfile
Comment (gavanderhoorn) [reply tesseract/Dockerfile]: On both Ubuntu 16.04 and Ubuntu 18.04, the following is true for me:

```shell
$ cat /etc/ld.so.conf.d/libc.conf
# libc default configuration
/usr/local/lib
$ dpkg -S /etc/ld.so.conf.d/libc.conf
libc-bin: /etc/ld.so.conf.d/libc.conf
```

The Docker image used here also seems to contain the same file:

```
$ docker run -it --rm ros:melodic-perception cat /etc/ld.so.conf.d/libc.conf
# libc default configuration
/usr/local/lib
```

In any case: did things not work without it?

In general `LD_LIBRARY_PATH` is not something which should be set unless really needed.

File: ros-core-nvidia/kinetic/Dockerfile
Comment (gavanderhoorn) [reply ros-core-nvidia/kinetic/Dockerfile]: No, not necessarily.

Is there any notification system which uses these addresses?

File: ros-core-nvidia/kinetic/Dockerfile
Comment (gavanderhoorn) [reply ros-core-nvidia/kinetic/Dockerfile]: I would recommend to not do this. See my earlier comment.

File: tesseract/Dockerfile
Comment (gavanderhoorn) [reply tesseract/Dockerfile]: No, it's ok to group `apt install`s of multiple packages. What I meant was: installing things right before they are used is beneficial as it can help Docker with reusing layers (the larger the set of packages you install, the higher the chance that you won't do that in any other `Dockerfile` with the exact same set -> no reuse of the layer).

So instead of installing all `.deb`s right at the beginning, you could install them in batches which make sense. I wouldn't go for separate lines for each package. That doesn't scale and is unmaintainable.

But again, not important here.

File: tesseract/Dockerfile
Comment (gavanderhoorn) [reply tesseract/Dockerfile]: > Are you recommending making them one line? 

you could, but if anything fails during the build, a restart of your `docker build` will need to `clone` everything again as well. That can quickly become rather annoying.

> If that is the case how do you leverage WORKDIR? 

Personally I would probably do something like this (but that's not really that different from what you are doing):

```Dockerfile
WORKDIR /bullet3/build
RUN git clone --branch master https://github.com/bulletphysics/bullet3.git /bullet3 \
 && cmake \
      -DBUILD_SHARED_LIBS=ON \
      -DBUILD_UNIT_TESTS=OFF \
      -DBULLET2_MULTITHREADING=OFF \
      -DCMAKE_BUILD_TYPE=Release \
      -DCMAKE_C_FLAGS="-w" \
      -DCMAKE_CXX_FLAGS="-w -std=c++14" \
      -DUSE_DOUBLE_PRECISION=ON \
    /bullet3 \
 && make -j8 \
 && make install \
 && cd / \
 && rm -rf /bullet3 /bullet3_build
```

This is the "old way" of doing things btw with CMake. Nowadays `cmake -S /bullet3 -B /bullet3_build` and then `cmake --build /bullet3_build` is 'sexier'. The build dir is created if it doesn't exist, which is nice.

File: tesseract/Dockerfile
Comment (Levi-Armstrong) [reply tesseract/Dockerfile]: I don't believe it was, which is why I think it was added. I can remove it for now and if it does not work I will create a separate PR adding it back.

Comment (gavanderhoorn) [reply tesseract/Dockerfile]: If it is needed, it would be an indication of something else not being correct.

File: ros-core-nvidia/kinetic/Dockerfile
Comment (Levi-Armstrong) [reply ros-core-nvidia/kinetic/Dockerfile]: I do not know. Can we just remove this line?

File: ros-core-nvidia/kinetic/Dockerfile
Comment (Levi-Armstrong) [reply ros-core-nvidia/kinetic/Dockerfile]: OK, I will go through and update the other.

File: tesseract/Dockerfile
Comment (Levi-Armstrong) [reply tesseract/Dockerfile]: I tried this but I could not get it to work immediately. I look at it when I have more time.

**Comment by Levi-Armstrong** [comment]
@gavanderhoorn How does it look now?

File: ros-core-nvidia/kinetic/Dockerfile
Comment (gavanderhoorn) [reply ros-core-nvidia/kinetic/Dockerfile]: Guess so.

File: tesseract/Dockerfile
Comment (gavanderhoorn) [reply tesseract/Dockerfile:29]: Not sure that's what containers are for, but OK.

**Review by gavanderhoorn — APPROVED** [review]
Looks ok.

**Comment by Levi-Armstrong** [comment]
> Not sure that's what containers are for, but OK.

What do you mean by this comment?

**Comment by haudren** [comment]
I think what @gavanderhoorn meant was that docker steps are supposed to be idempotent, e.g. the result of a step should only depend on the filesystem state. In your case, rebuilding the docker container will pull the latest bullet code if (and only if) one of the previous steps has changed, regardless of the actual state of the bullet repository.

Typically, I clone the repository in my docker context, then use the `COPY` instruction to bring the files in. This makes sure to trigger a rebuild if the file content has changed.

**Pull request merged**
